### ISSUE TITLE
nginx - how to?

### ISSUE THREAD

**Issue by gvenka008c** [nginxinc/kubernetes-ingress#130]
I have installed nginx on a separate server outside of kube cluster. I have installed nginx ingress controller. How to I configure nginx controller to talk to nginx server and update the nginx templates dynamically? Also how can i update multiple nginx servers?

**Comment by pleshakov** [comment]
@gvenka008c 

The controller is recommended to be deployed in a pod inside the cluster. Please help me understand why you would like to have the controller deployed outside of the cluster.

**Comment by gvenka008c** [comment]
@pleshakov the ingress controller is deployed inside the k8 clusters. I want to install nginx on a separate vm and update nginx config with pod details for load balancing.

**Comment by pleshakov** [comment]
@gvenka008c thanks for the details.

In this case you can deploy the controller on all or a subset of nodes of the cluster using the [DaemonSet example](https://github.com/nginxinc/kubernetes-ingress/tree/master/examples/daemon-set) and then configure NGINX to load balance among those nodes on exposed ports. Will that work for you?

**Comment by gvenka008c** [comment]
@pleshakov Yes, that's what I am currently looking for. How does nginx controller connects to my Nginix load balance and update the nodes/exposed pods dynamically?

**Comment by hardymansen** [comment]
Im looking into the same, since i don't want too install nginx in a container. Because of port persistence during upgrades. My solution is probably to have a simple ugly internal host only REST service that can be triggered by nginx-controller on changes to do a nginx -s reload.

**Comment by pleshakov** [comment]
@gvenka008c @hardymansen 
Unfortunately, deploying NGINX separately from the Ingress controller is not supported. NGINX must be deployed with the controller in the same container. There is no ready solution to separate the controller from NGINX. 

Technically, you can run NGINX with the controller on a host. However, the following must be addressed:
- the controller must be properly authenticated with the Kubernetes API
- the host must have access to the internal overlay network of the cluster so that NGINX can connect to containers.

What are the reasons that prevents you to deploy NGINX in a pod?

**Comment by hardymansen** [comment]
In our scenario nginx is used as the edge nodes to customers. AFAIK, to be able to upgrade nginx without affecting customer traffic i have to use nginx inmemory upgrade option and always have listener ports instead of switching containers. 
Also performance is key in our scenario and we want to maximize the bare metal. I don't want ex a firewall state machine on the edge that can be used for ex dos attacks.

**Comment by gvenka008c** [comment]
@pleshakov 

I deployed one of icinga2 application into kube cluster. 

```
# kubectl expose deployment icinga2 --type=NodePort --name=icinga2
service "icinga2” exposed
```

```
#  kubectl get services | grep icinga2
icinga2                  10.254.164.254   <nodes>       80:31706/TCP                                   24d
```

```
kubectl get pods -o wide | grep icinga2
icinga2-1541389751-fxdq3         1/1       Running   0          13d       172.30.17.2   xx.xxx.x.xx
```

```
 more icinga2-ingress.yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: icinga2
spec:
  rules:
  - host: icinga2
    http:
      paths:
      - path: /icingaweb2/*
        backend:
          serviceName: icinga2
          servicePort: 80
```

<img width="1163" alt="screen shot 2017-04-05 at 10 26 49 am" src="https://cloud.githubusercontent.com/assets/13439779/24710198/6958cf4a-19ea-11e7-8b5e-db663cf70bcd.png">


```
# more /etc/nginx/default-icinga2.conf

upstream default-icinga2-icinga2-icinga2 {

	server 172.30.17.2:80;
}


server {
	listen 80;

	server_name icinga2;


        location /icingaweb2/* {
                proxy_http_version 1.1;

                proxy_connect_timeout 60s;
                proxy_read_timeout 60s;
                client_max_body_size 1m;
                proxy_set_header Host $host;
                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_set_header X-Forwarded-Host $host;
                proxy_set_header X-Forwarded-Port $server_port;
                proxy_set_header X-Forwarded-Proto $scheme;

                proxy_buffering on;

                proxy_pass http://default-icinga2-icinga2-icinga2;

        }
}
```

When I hit http://<ingress-controller-ip>>:80/icingaweb2/ , it's not directing me to the icinga2 login page. Thoughts?

**Comment by pleshakov** [comment]
@gvenka008c 
Could you change the path in your Ingress resource from`/icingaweb2/*` to `/icingaweb2/`?
Note that regular expressions are not supported by the controller. However, for this case a regular expression  is not required.

**Comment by pleshakov** [comment]
@hardymansen 
Thanks for the details. In your network environment, does an NGINX machine have access to the overlay network of the cluster, so that NGINX can directly connect to containers?

**Comment by hardymansen** [comment]
Right now this is only on planning stage, but yes the plan is to use something like this http://www.bigswitch.com/videos/kubernetes-container-orchestration-with-big-cloud-fabric

**Comment by pleshakov** [comment]
@hardymansen In this case, to update NGINX configuration you can run the controller along with NGINX:

1.  Compile the controller using the provided Makefile: `$ make nginx-ingress`
1.  Run the controller as a system service.

The controller requires root user for reloading NGINX and writing NGINX configuration files and SSL certificates in NGINX directories (/etc/nginx) Also, the controller must be properly authenticated with the Kubernetes API. `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` env variables must point to the Kubernetes API server and `/var/run/secrets/kubernetes.io/serviceaccount/` must contain the certificate (ca.crt) and the token for authentication with the API server.

**Comment by gvenka008c** [comment]
@pleshakov I made the change as you suggested but still says 404 Not Found

```
# more default-icinga2.conf

upstream default-icinga2-icinga2-icinga2 {

	server 172.30.17.2:80;
}


server {
	listen 80;

	server_name icinga2;

        location /icingaweb2/ {
                proxy_http_version 1.1;

                proxy_connect_timeout 60s;
                proxy_read_timeout 60s;
                client_max_body_size 1m;
                proxy_set_header Host $host;
                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_set_header X-Forwarded-Host $host;
                proxy_set_header X-Forwarded-Port $server_port;
                proxy_set_header X-Forwarded-Proto $scheme;

                proxy_buffering on;

                proxy_pass http://default-icinga2-icinga2-icinga2;

        }
}
```

```
curl http://xx.xxx.x.xxx/icingaweb2/
<html>
<head><title>301 Moved Permanently</title></head>
<body bgcolor="white">
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx/1.11.8</center>
</body>
</html>
```

**Comment by pleshakov** [comment]
@gvenka008c 
Please change `/icingaweb2/` to `/icingaweb2` in your Ingress rule.

**Comment by gvenka008c** [comment]
@pleshakov Still the same. No luck :(

```
more default-icinga2.conf

upstream default-icinga2-icinga2-icinga2 {

	server 172.30.17.2:80;
}


server {
	listen 80;


	server_name icinga2;

        location /icingaweb2 {
                proxy_http_version 1.1;

                proxy_connect_timeout 60s;
                proxy_read_timeout 60s;
                client_max_body_size 1m;
                proxy_set_header Host $host;
                proxy_set_header X-Real-IP $remote_addr;
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_set_header X-Forwarded-Host $host;
                proxy_set_header X-Forwarded-Port $server_port;
                proxy_set_header X-Forwarded-Proto $scheme;

                proxy_buffering on;

                proxy_pass http://default-icinga2-icinga2-icinga2;

        }
}
```

```
curl -IS http://xx.xxx.x.xxx:80/icingaweb2
HTTP/1.1 301 Moved Permanently
Server: nginx/1.11.8
Date: Thu, 06 Apr 2017 20:45:26 GMT
Content-Type: text/html
Content-Length: 185
Connection: keep-alive
Location: https://xx.xxx.x.xxx/icingaweb2
```

**Comment by pleshakov** [comment]
@gvenka008c 
It looks like the backend sends a redirect to https. You can configure ssl termination in your Ingress resource, so that NGINX will be configured to handle https after the redirect.

**Comment by gvenka008c** [comment]
@pleshakov is that part of ingress resource creation or has to be at ingress controller changes? Got any example that you can suggest?

**Comment by hardymansen** [comment]
@pleshakov Thank you. Yes i also came to the same conclusion. I will try running it as a system service with connection into the cluster and see if it can work in our hybrid environment. Think it can be a good match, will be interesting.

**Comment by pleshakov** [comment]
@gvenka008c 
SSL termination is part of Ingress. See the cafe-example.

It looks like the backend returned 404.  To make sure, please check NGINX logs through the `kubectl logs` command. Note that your backend must be able to handle `/icingaweb2/authentication` URI.

NGINX connects to backend containers directly through the overlay network and doesn't rely on NodePorts.

**Comment by notxcain** [comment]
@hardymansen how was your experience of running nginx ingress outside of k8s?

**Comment by Mazon** [comment]
I have switched to https://github.com/kubernetes/ingress the contrib version of ingress controller, because that one seem to be more up2date, when it comes to to using k8 api etc. 
It is working great! 

This is what you have to do:
- Make sure your edge load balancer can reach your pods and api server. Using example calico bgp/bird to have a visible network.
- Fix so ingress controller don't try to start nginx or opens ports. Im working on a nicer PR (or fork, but dont want that. If upstream want it)

Basically conditional on Start() in controllers/nginx/pkg/cmd/controller/nginx.go and the listen directive in newNGINXController() in same file. 

Then it works.

**Comment by hardymansen** [comment]
I have switched to https://github.com/kubernetes/ingress the contrib version of ingress controller, because that one seem to be more up2date, when it comes to to using k8 api client libs etc.
It is working great!

This is what you have to do:

    Make sure your edge load balancer can reach your pods and api server. Using example calico bgp/bird to have a visible network.
    Fix so ingress controller don't try to start nginx or opens ports. Im working on a nicer PR (or fork, but dont want that. If upstream want it)

Basically conditional on Start() in controllers/nginx/pkg/cmd/controller/nginx.go and the listen directive in newNGINXController() in same file.

Then it works. Only thing ive seen sofar is the status update is not setting the correct values back to the ingress object. Will see if i can get some automation to pickup values from hosts ip stack instead of trying k8 stuff.

PS. I am creating a separate backend config for external nginx , so will be /controllers/ext_nginx
and maybe it will be my github fork or upstream, will have it polished in a few days. will update link

**Comment by hardymansen** [comment]
You have the branch here: https://github.com/unibet/ingress/tree/extnginx , it should work. (tm)
